### PR TITLE
Feature : Implement a AWS Session manager Parameter Store Operators and hooks

### PR DESCRIPTION
**Pull request by hsrocks**
**related: #21008 and #21012**

This change is to implement the operators and hooks for AWS Session manager parameter store to get the value of single parameter , multiple parameters by path, and to add new parameter and its value to parameter store

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by hsrocks** [comment]
Updated the operators and hooks for SSM as per latest code changes

File: airflow/providers/amazon/aws/hooks/ssm_parameter_store.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        except ssm_client.exceptions.ParameterNotFound:
+            self.log.error('Parameter Does Not Exist')
+            raise
+        except Exception as general_error:
+            self.log.error("Failed to list SSM Parameter, error: %s", general_error)
+            raise
+
+    def get_parameters_by_path(self, path: str, **parameter_kwargs):
+        """
+        Retrieve parameter value from AWS Systems Manager Parameter Store.
```
Comment (subkanthi) [inline airflow/providers/amazon/aws/hooks/ssm_parameter_store.py]: This looks like a copy/paste error, should this be retrieve parameter value by path maybe?

File: airflow/providers/amazon/aws/hooks/ssm_parameter_store.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        try:
+            get_parameter_by_path_response = ssm_client.get_parameters_by_path(
+                Path=path,
+                **parameter_kwargs,
+            )
+            if 'Parameters' in get_parameter_by_path_response:
+                self.log.info("Returning dictionary with the list of parameters")
+                return get_parameter_by_path_response
+        except Exception as general_error:
+            self.log.error("Failed to list SSM Parameter, error: %s", general_error)
```
Comment (subkanthi) [inline airflow/providers/amazon/aws/hooks/ssm_parameter_store.py:81]: Line 59 and Line 81 has the same error message, maybe something different to distinguish.

File: airflow/providers/amazon/aws/hooks/ssm_parameter_store.py (hunk 1 above)
Comment (hsrocks) [reply airflow/providers/amazon/aws/hooks/ssm_parameter_store.py]: Thanks @subkanthi . Fixed it

**Comment by potiuk** [comment]
Also fixing static checks alongside the rebase would be great.

**Comment by hsrocks** [comment]
> Also fixing static checks alongside the rebase would be great.

Will do . Can see static check failure is not related to SSM changes but will relook it again and try to fix it. Thanks :)

**Comment by dbarrundiag** [comment]
I would be curious to learn how these hooks and operators are related to the already existing logic available in Systems Manager Backend in Amazon Provider package: `airflow.providers.amazon.aws.secrets.systems_manager.SystemsManagerParameterStoreBackend` 

I wonder if we should leverage these hooks in the Backend class now? it seems like some code is repeated in these 2?

CC: @potiuk @o-nikolas

**Comment by hsrocks** [comment]
> I would be curious to learn how these hooks and operators are related to the already existing logic available in Systems Manager Backend in Amazon Provider package: `airflow.providers.amazon.aws.secrets.systems_manager.SystemsManagerParameterStoreBackend`
> 
> I wonder if we should leverage these hooks in the Backend class now? it seems like some code is repeated in these 2?
> 
> CC: @potiuk @o-nikolas

Thanks for review @dbarrundiag !! While I am not sure on the reason but I think they want to ensure segregation between hooks/operators and secrets module. But if thats not true the we can use secrets_manager hook in this class as well: airflow.providers.amazon.aws.secrets.secrets_manager.SecretsManagerBackend

**Comment by hsrocks** [comment]
@potiuk @o-nikolas Can you please review the request

**Comment by o-nikolas** [comment]
> @potiuk @o-nikolas Can you please review the request

Hey folks,
Sorry, it's been on my ToDo list to dive into this PR, but I just haven't had the time. But at a high level, re: the earlier discussion,   I see no reason on our end why their should be any duplicated code. If we have a new hook that provides the same functionality needed by the manager backend, we should reuse that code.

**Comment by potiuk** [comment]
hey @hsrocks - sorry for the delay, I think you should take a look how Hashicorp Vault is implemented. It has very similar story. And what we came up with is:

Internal Client that is reused by both Hook and Secret Manager:
* https://github.com/apache/airflow/blob/main/airflow/providers/hashicorp/_internal_client/vault_client.py

Hook (which is a thin wrapper over the client):
* https://github.com/apache/airflow/blob/main/airflow/providers/hashicorp/hooks/vault.py

Secrets Manager (which is a thin wrapper over the client):
* https://github.com/apache/airflow/blob/main/airflow/providers/hashicorp/secrets/vault.py

Duplication is minimized (and we've already implemented some features in client that were automatically available to both SecretManager and Hook (see the history of "vault_client.py" - proving that this pattern is really good.

**Pull request closed without merging**